CondFormats/HcalObjects/src/HcalElectronicsMap.cc

```diff
@@ -16,25 +16,57 @@ POOL object to store mapping for Hcal channels
 HcalElectronicsMap::HcalElectronicsMap() : 
   mPItems(HcalElectronicsId::maxLinearIndex+1),
   mTItems(HcalElectronicsId::maxLinearIndex+1),
-  sortedByPId(false),
-  sortedByTId(false)
+  mPItemsById(nullptr), mTItemsByTrigId(nullptr)
 {}
 
 namespace hcal_impl {
   class LessById {public: bool operator () (const HcalElectronicsMap::PrecisionItem* a, const HcalElectronicsMap::PrecisionItem* b) {return a->mId < b->mId;}};
   class LessByTrigId {public: bool operator () (const HcalElectronicsMap::TriggerItem* a, const HcalElectronicsMap::TriggerItem* b) {return a->mTrigId < b->mTrigId;}};
 }
 
-HcalElectronicsMap::~HcalElectronicsMap(){}
+HcalElectronicsMap::~HcalElectronicsMap() {
+    delete mPItemsById;
+    mPItemsById = nullptr;
+    delete mTItemsByTrigId;
+    mTItemsByTrigId = nullptr;
+}
+// copy-ctor
+HcalElectronicsMap::HcalElectronicsMap(const HcalElectronicsMap& src)
+    : mPItems(src.mPItems), mTItems(src.mTItems),
+      mPItemsById(nullptr), mTItemsByTrigId(nullptr) {}
+// copy assignment operator
+HcalElectronicsMap&
+HcalElectronicsMap::operator=(const HcalElectronicsMap& rhs) {
+    HcalElectronicsMap temp(rhs);
+    temp.swap(*this);
+    return *this;
+}
+// public swap function
+void HcalElectronicsMap::swap(HcalElectronicsMap& other) {
+    std::swap(mPItems, other.mPItems);
+    std::swap(mTItems, other.mTItems);
+    other.mTItemsByTrigId.exchange(
+            mTItemsByTrigId.exchange(other.mTItemsByTrigId.load(std::memory_order_acquire), std::memory_order_acq_rel),
+            std::memory_order_acq_rel);
+    other.mPItemsById.exchange(
+            mPItemsById.exchange(other.mPItemsById.load(std::memory_order_acquire), std::memory_order_acq_rel),
+            std::memory_order_acq_rel);
+}
+// move constructor
+HcalElectronicsMap::HcalElectronicsMap(HcalElectronicsMap&& other) 
+    : HcalElectronicsMap() {
+    other.swap(*this);
+}
 
 const HcalElectronicsMap::PrecisionItem* HcalElectronicsMap::findById (unsigned long fId) const {
   PrecisionItem target (fId, 0);
   std::vector<const HcalElectronicsMap::PrecisionItem*>::const_iterator item;
 
-  if (!sortedByPId) sortById();
+  sortById();
   
-  item = std::lower_bound (mPItemsById.begin(), mPItemsById.end(), &target, hcal_impl::LessById());
-  if (item == mPItemsById.end() || (*item)->mId != fId) 
+  auto ptr = (*mPItemsById.load(std::memory_order_acquire));
```

Comment:
as a part of testing #15321 which somewhat excessively uses the HcalElectronicsMap,
I noticed that the CPU use is dominated by memmove
https://slava77sk.web.cern.ch/slava77sk/reco/cgi-bin/igprof-navigator/CMSSW_8_1_0_pre10-sign746-ccb128f.tt35pre8.mt8.pp/947
After changing to ```auto const& ptr``` the time in hcalRawData went down from ~2s/evt to ~40 ms/evt.

@lihux25 or @igv4321 please follow up